src/models/document/base-document-content.ts

```diff
@@ -1002,215 +1004,153 @@ export const BaseDocumentContentModel = types
-                  columnWidths[attributeIdMap[originalId]] = width;
-              }
-            }
-          });
-          tileContent.content = { ...oldContent.content, columnWidths };
-        } else if (tile.tileType === "DataCard") {
-          const oldAttributeId = oldContent.content.selectedSortAttributeId;
-          sharedDataSetEntries.forEach(sharedDataSetEntry => {
-            const originalSharedDataSetId = sharedDataSetEntry.sharedModel.id;
-            const attributeIdMap = updatedSharedModelMap[originalSharedDataSetId].attributeIdMap;
-            if (attributeIdMap[oldAttributeId]) {
-              const content = { ...oldContent.content, selectedSortAttributeId: attributeIdMap[oldAttributeId] };
-              tileContent.content = content;
-            }
-          });
+        const updateFunction = getTileContentInfo(tile.tileType)?.updateContentWithNewSharedModelIds;
+        if (updateFunction) {
+          tileContent.content = updateFunction(oldContent.content, tileSharedModelEntries, updatedSharedModelMap);
         }
 
         // Save the updated tile so we can add it to the document
-        updatedTiles.push({
-          ...tile,
-          newTileId: tileIdMap[tile.tileId],
-          tileContent: JSON.stringify(tileContent)
-        });
+        updatedTiles.push({ ...tile, newTileId: tileIdMap[tile.tileId], tileContent: JSON.stringify(tileContent) });
       });
 
       // Add copied tiles to document
-      const results = self.copyTilesIntoNewRows(updatedTiles, rowIndex);
+      const results = insertTileFunction(updatedTiles, rowInfo);
 
       // Increment default titles when necessary
       results.forEach((result, i) => {
-        const newTile = result?.tileId && self.getTile(result.tileId);
-        if (result && newTile) {
-          const tileContentInfo = getTileContentInfo(newTile.content.type);
-          if (tileContentInfo) {
-            const match = titleMatchesDefault(newTile.title, tileContentInfo.titleBase);
-            if (match) {
-              const newTitle = self.getNewTileTitle(newTile.content.type);
+        if (result?.tileId) {
+          const { oldTitle, newTitle } = self.updateDefaultTileTitle(result.tileId);
 
-              // If the tile title needed to be update, we assume we should also update the data set's name
-              newSharedModelEntries.forEach(sharedModelEntry => {
-                const oldName = sharedModelEntry.sharedModel.dataSet.name;
-                if (oldName === newTile.title) {
-                  sharedModelEntry.sharedModel.dataSet.name = newTitle;
+          // If the tile title needed to be updated, we assume we should also update the data set's name
+          if (newTitle && sharedModelEntries) {
+            newSharedModelEntries.forEach(sharedModelEntry => {
+              if (isSharedDataSetSnapshot(sharedModelEntry.sharedModel)) {
+                const sharedDataSet = sharedModelEntry.sharedModel as SharedDataSetSnapshotType;
```

Comment:
```suggestion
                const sharedDataSet = sharedModelEntry.sharedModel;
```
The cast should no longer be necessary.